src/lib.rs

```diff
@@ -264,23 +267,24 @@ async fn splice_impl(
     let fd_in = asyncfd_in.as_raw_fd();
     let fd_out = asyncfd_out.as_raw_fd();
 
-    let off_in = as_ptr(off_in);
-    let off_out = as_ptr(off_out);
-
     let flags = libc::SPLICE_F_NONBLOCK
         | if has_more_data {
             libc::SPLICE_F_MORE
         } else {
             0
         };
 
-    read_ready.clear_ready();
-    write_ready.clear_ready();
-
     loop {
-        let ret = unsafe { libc::splice(fd_in, off_in, fd_out, off_out, len, flags) };
+        let ret =
+            unsafe { libc::splice(fd_in, as_ptr(off_in), fd_out, as_ptr(off_out), len, flags) };
         match cvt!(ret) {
-            Err(e) if is_wouldblock(&e) => (),
+            Err(e) if is_wouldblock(&e) => {
+                read_ready.clear_ready();
+                write_ready.clear_ready();
```

Comment:
How do we know which end blocked?

Reply:
Honestly, I don't know the correct way here.

I don't know for sure whether it is reader that is going to block or writer.

Maybe I should just launch a `libc::read` and `libc::write` to test it out.

At this rate, I might need to completely rewrite the `splice` and `tee` API in another PR.

Reply:
Aargh, the libc splice API doesn't tell which one blocked :(
But maybe that's actually fine... Tokio will poll this future again once any readiness notification comes from any of those two. So that's probably correct.

Reply:
Aha, sorry I was distracted right now, I will come back to this later.

Reply:
> But maybe that's actually fine... Tokio will poll this future again once any readiness notification comes from any of those two. So that's probably correct. But then you probably don't want to wait on readable / writable in 213, 214, but just go straight to the splice call.

Hmmm...

Let me have a try.

Reply:
After a second thought, I think just polling reader or writer is not robust, so I decide to go with the original idea of manually `read` and `write` to test whether they are ready.

Reply:
@pkolaczk Done!

Can you please try this again?

Reply:
Data gets through this time, but it spins on a core again after the first message.

Reply:
The read end is stuck marked readable after reading the first message, even though there is not more data available.

Reply:
No, it is not because of that select.
First, that select was there before my attempt to switch to zero copy and it worked flawlessly.
Second, if you look carefully, the code doing splicing is enclosed in a loop, so that select has totally nothing to do with it - the future in the select does not complete until the stream is closed.

Reply:
Anyway, I removed the select totally (assuming I'm testing only one half of the connection now), and the problem remains.

Reply:
```
        src.readable().await?;                        // this passes immediately on the second loop cycle, even though there are no data
```

This is because `splice_impl` does not clear the ready flag on success.
But this should not become a problem, since `test_read_ready` should handle that situation.

Possible reasons for infinite loop is that `splice_impl` and `tee_impl` does not handle the case where Eof happens and `splice`/`tee` returns 0.

Reply:
Wait, but I didn't even get to the EOR. The stream is still open when the spin happens.

Reply:
Can you use the debugger or `strace` to find out which line the code is blocked on?

Reply:
Spinning on splice / read / write:
```
[pid 50938] read(7, "", 0)              = 0
[pid 50938] write(12, "", 0)            = 0
[pid 50938] splice(7, NULL, 12, NULL, 4096, SPLICE_F_NONBLOCK) = -1 EAGAIN (Zasoby chwilowo niedostępne)
[pid 50938] read(7, "", 0)              = 0
[pid 50938] write(12, "", 0)            = 0
[pid 50938] splice(7, NULL, 12, NULL, 4096, SPLICE_F_NONBLOCK) = -1 EAGAIN (Zasoby chwilowo niedostępne)
[pid 50938] read(7, "", 0)              = 0
[pid 50938] write(12, "", 0)            = 0
[pid 50938] splice(7, NULL, 12, NULL, 4096, SPLICE_F_NONBLOCK) = -1 EAGAIN (Zasoby chwilowo niedostępne)
[pid 50938] read(7, "", 0)              = 0
[pid 50938] write(12, "", 0)            = 0
[pid 50938] splice(7, NULL, 12, NULL, 4096, SPLICE_F_NONBLOCK) = -1 EAGAIN (Zasoby chwilowo niedostępne)
[pid 50938] read(7, "", 0)              = 0
[pid 50938] write(12, "", 0)            = 0
[pid 50938] splice(7, NULL, 12, NULL, 4096, SPLICE_F_NONBLOCK) = -1 EAGAIN (Zasoby chwilowo niedostępne)
[pid 50938] read(7, "", 0)              = 0
[pid 50938] write(12, "", 0)            = 0
[pid 50938] splice(7, NULL, 12, NULL, 4096, SPLICE_F_NONBLOCK) = -1 EAGAIN (Zasoby chwilowo niedostępne)
[pid 50938] read(7, "", 0)              = 0
[pid 50938] write(12, "", 0)            = 0
[pid 50938] splice(7, NULL, 12, NULL, 4096, SPLICE_F_NONBLOCK) = -1 EAGAIN (Zasoby chwilowo niedostępne)
[pid 50938] read(7, "", 0)              = 0
[pid 50938] write(12, "", 0)            = 0
[pid 50938] splice(7, NULL, 12, NULL, 4096, SPLICE_F_NONBLOCK) = -1 EAGAIN (Zasoby chwilowo niedostępne)
[pid 50938] read(7, "", 0)              = 0
[pid 50938] write(12, "", 0)            = 0
[pid 50938] splice(7, NULL, 12, NULL, 4096, SPLICE_F_NONBLOCK) = -1 EAGAIN (Zasoby chwilowo niedostępne)
```

Reply:
And logically we're blocked in the `splice_from` future. So tokio is polling that future all the time as fast as it can.

Reply:
Yeah, read/write return 0, success, so it doesn't clear the readable/writable readiness.

Reply:
I will switch to `select` then...
It's efficient enough for polling two fds.

Reply:
I just wonder, what happens if we clear readiness of both? We cannot be reading / writing to those FDs anyway, when we are logically still running that future. So maybe we are overthinking it.

Reply:
I remembered that `tokio` uses `epoll` with the edge-triggered mode, meaning once `epoll` considers a fd as ready, it won't return readiness for it again until the syscalls performed on it returns `EAGAIN`.

Reply:
There's a detailed description of edge-triggered vs level-triggered in the [man page](https://man7.org/linux/man-pages/man7/epoll.7.html).

Reply:
I mean - If we clear both of them upon receiving EAGAIN, one of them will trigger the epoll event and will make tokio poll us again. Then we call splice and if it succeeds, we set readable / writable status on both again.

Reply:
> one of them will trigger the epoll event and will make tokio poll us again

Yeah, but then we would block forever when `epoll`ing the other one, who doesn't return `EAGAIN`.

Reply:
But who/what would block on epolling the other one?

Reply:
I've pushed another commit that attempts to fix this using `poll`.

`select`'s API is too hard to use, so I decided to use `poll` instead.

Reply:
> one of them will trigger the epoll event and will make tokio poll us again.

You misunderstood how it works.

`epoll` treats each fd separately.
So I/O on one fd returns `EAGAIN` under edge-triggered mode makes `epoll` poll for that particular fd again, but the other fd does not get polled.

Reply:
Ok. Now it works, the communicatin goes both ways and it doesn't spin. However, I notice there is a visible, quite large latency. Responses from an echo server through this proxy come after 1 second on loopback.

Reply:
Ah, got it. Setting has_more_data to `false` reduced the delay.

Reply:
> has_more_data to `false`

Oops, I didn't know that.
I guess the kernel was doing some preparation for more data, like allocation?

How much of the latency do you get after disable `has_more_data`?

Reply:
Out of curiosity, how much of the throughput do you get and how does it compare to simple read + write?

Reply:
I don't have a good framework for measuring that accurately (yet), but it is a night-and-day difference. 
Like, with `has_more_data = true` it felt as if I was talking to a machine on the other side of the planet. 
After setting to false, responses are immediate, I can't sense I'm interacting through a proxy.

I guess it just waits for more data to come, before pushing out a packet to the remote end.

Reply:
> I guess it just waits for more data to come, before pushing out a packet to the remote end.

Emmm, that's completely unexpected...

Reply:
> Out of curiosity, how much of the throughput do you get and how does it compare to simple read + write?

I managed to measure it once, and got 1.4 GB/s when pushing data from `/dev/zero` -> `nc`  -> my proxy -> `nc` -> /dev/null.
But with traditional copying I'm getting 1.3 GB/s so honestly it is a bit disappointing. 

Tried to measure it more than once, but it is constantly locking up. Something is still not ok.